File: Data Structures/Stacks & Queues/queue.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

// FIFO
class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.length = 0;
  }

  /**
   * @returns the first element of the queue
   */
  peek() {
    return this.first;
  }

  /**
   * add item to the end of the queue
   */
  enqueue(item) {
    const newNode = new Node(item);

    if (this.length === 0) {
      this.first = newNode;
      this.last = newNode;
    } else {
      this.last.next = newNode;
      this.last = newNode;
    }

    this.length++;
  }

  /**
   * Removes the first element of the queue
   * @returns the removed element
   */
  dequeue() {
    if (this.length === 0) {
      return null;
    }

    if (this.length === 1) {
      this.last = null;
    }

    const holdingPointer = this.first;
    this.first = this.first.next;

    this.length--;

    return holdingPointer;
  }
}

const myQueue = new Queue();
myQueue.enqueue("1");
myQueue.enqueue("2");
myQueue.enqueue("3");
myQueue.enqueue("4");
myQueue.enqueue("5");
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();

console.clear();
console.dir(myQueue, { depth: 10 });

File: Data Structures/Stacks & Queues/queue.py
# Queue - FIFO
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.first = None
        self.queue = None
        self.length = 0

    # returns the first element of the queue
    def peek(self):
        return self.first

    # add element to the end of the queue
    def enqueue(self, item):
        new_node = Node(item)

        if self.length == 0:
            self.first = new_node
            self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node

        self.length += 1

    # remove the first element of the queue and then returns it
    def dequeue(self):
        if self.length == 0:
            return None

        if self.length == 1:
            self.last = None

        holding_pointer = self.first
        self.first = self.first.next

        self.length -= 1
        return holding_pointer


my_queue = Queue()
my_queue.enqueue(1)
my_queue.enqueue(2)
my_queue.enqueue(3)
my_queue.enqueue(4)
my_queue.enqueue(5)
my_queue.enqueue(6)
my_queue.dequeue()
my_queue.dequeue()
my_queue.dequeue()
my_queue.dequeue()

print(my_queue.first, my_queue.last, my_queue.length)
